sdk/dotnet/examples/bucket/TfGenStyleResources.cs

```diff
@@ -0,0 +1,61 @@
+// These are some hand authored resources in the style of what we think we'd generate via `tfgen`.  So we'll get
+// the shape right "by hand" and then work on the code-gen to stub everything else out:
+
+using Pulumi;
+using Pulumi.Rpc;
+
+namespace AWS.S3
+{
+    public class BucketObject : CustomResource
+    {
+        public BucketObject(string name, BucketObjectArgs args, ResourceOptions opts = null)
+            : base("aws:s3/bucketObject:BucketObject", name, args, opts)
+        {
+            OnConstructorCompleted();
+        }
+    }
+
+    public class BucketObjectArgs : ResourceArgs
+    {
+        public Input<string> Acl;
+        public Input<Id> Bucket;
+        public Input<string> ContentBase64;
+        public Input<string> ContentType;
+        public Input<string> Key;
+        public Input<AssetOrArchive> Source;
+
+        protected override void AddProperties(PropertyBuilder builder)
+        {
+            builder.Add("acl", Acl);
+            builder.Add("bucket", Bucket);
+            builder.Add("contentBase64", ContentBase64);
+            builder.Add("contentType", ContentType);
+            builder.Add("key", Key);
+            builder.Add("source", Source);

```

Comment:
How do I pass complex objects? This works but looks complicated:
``` cs
    public class PlanSkuArgs
    {
        public Input<string> Tier { get; set; }
        public Input<string> Size { get; set; }
    }

public class PlanArgs : ResourceArgs
    {
        public Input<string> Kind { get; set; }
        public Input<string> Location { get; set; }
        public Input<string> ResourceGroupName { get; set; }
        public Input<PlanSkuArgs> Sku { get; set; }

        protected override void AddProperties(PropertyBuilder builder)
        {
            builder.Add("kind", Kind);
            builder.Add("location", Location);
            builder.Add("resourceGroupName", ResourceGroupName);

            Input<Dictionary<string, Input<string>>> dict = 
                Sku.ToOutput()
                .Apply(sku => 
                    new Dictionary<string, Input<string>>
                    {
                        { "tier",  sku.Tier },
                        { "size",  sku.Size },
                    });
            builder.Add("sku", dict);
        }
    }
```

Reply:
> How do I pass complex objects? This works but looks complicated:

You would just need to do this:

```c#
        protected override void AddProperties(PropertyBuilder builder)
        {
            builder.Add("kind", Kind);
            builder.Add("location", Location);
            builder.Add("resourceGroupName", ResourceGroupName);
            builder.Add("sku", sku);
        }
```

The underlying layer is designed to handle ResourceArgs deep down (it will recursively call ToDictionary on them).

Reply:
This gives
```
System.InvalidOperationException: Pulumi.Azure.AppService.PlanSkuArgs is not a supported argument type.
        resource:asp[azure:appservice/plan:Plan].sku.id
       at Pulumi.Deployment.SerializePropertyAsync(String ctx, Object prop, HashSet`1 dependentResources) in C:\work\pulumi\dotnet\pulumi\sdk\dotnet\Pulumi\Runtime\Deployment_Serialization.cs:line 280
       at Pulumi.Deployment.SerializePropertyAsync(String ctx, Object prop, HashSet`1 dependentResources) in C:\work\pulumi\dotnet\pulumi\sdk\dotnet\Pulumi\Runtime\Deployment_Serialization.cs:line 171
       at Pulumi.Deployment.SerializePropertyAsync(String ctx, Object prop, HashSet`1 dependentResources) in C:\work\pulumi\dotnet\pulumi\sdk\dotnet\Pulumi\Runtime\Deployment_Serialization.cs:line 153
```

Does `PlanSkuArgs` have to implement `ResourceArgs`?

Reply:
yes it does.

Reply:
clear then, thanks!